BLD: switch to meson-python as the default build backend

rgommers opened the pull request:
Time to give this a go now we're developing for 2.0 on `main`. Will also make it easier to use NumPy `main` with Python 3.12 dev builds. 

Follows up on https://github.com/numpy/numpy/issues/23808#issuecomment-1566872422.

rgommers committed:
BLD: default to using meson-python as build backend

rgommers committed:
CI: tweak CI for the change to meson-python as default build backend

rgommers committed:
CI: drop 32-bit Windows Azure testing, adjust 64-bit for Meson

Closes gh-23717

rgommers committed:
BLD: fix bug with CMake fallback detection of BLAS/LAPACK

rgommers committed:
CI: upgrade rtools to 4.3 to get a new enough GCC version in Azure

The rtools 4.0 used GCC 8.3 and we set the minimum to 8.3 for the Meson
build. That was to avoid workarounds for SIMD support for old GCCs,
and seems to still be a reasonable lower bound.

rgommers committed:
CI: keep the Emscripten/Pyodide job on a setup.py-based build

charris commented:
I think recent versions of rtools are preinstalled in GitHub actions. I suspect it is possible to split windows builds into separate cases for 32 and 64 bits and use the appropriate version of rtools. Of course, we could also just drop 32 bit builds for 2.0.

rgommers commented:
> Of course, we could also just drop 32 bit builds for 2.0.

This one:) xref gh-23717.

mattip commented:
Another option is to use MSVC on windows. We need 2 things from rtools: mingw (for gcc) and gfortran. We used to require mingw in order to mangle the openblas.a file to create an import library, but no longer need to do that, so mingw is no longer required. gfortran is needed for fortran, but that is only needed in the f2py tests. We could skip them on win32 (if gforfrtran is not found).

rgommers commented:
A few different failures left:

```
..\..\meson.build:30:4: ERROR: Problem encountered: NumPy requires GCC >= 8.4
```
Should not be too difficult to resolve, we need the right version from Rtools. @mattip will try to fix that up next week once he has access to his Windows machine.

```
FAILED ..\build-install\Lib\site-packages\numpy\tests\test_public_api.py::test_array_api_entry_point
```
The `.dist-info` directory of the wheel does seem to contain the right `entry_points.txt` entry, but the entry point doesn't get installed for some reason. I'll look into this. 

```
numpy/core/_multiarray_umath.cpython-310-aarch64-linux-gnu.so: undefined symbol: cblas_cdotc_sub
```
This is due to the wheel being built against ILP64 OpenBLAS, and the symbol suffix handling for that needs wiring up.

mattip commented:
meson-python [can handle setting up windows](https://meson-python.readthedocs.io/en/latest/how-to-guides/meson-args.html#force-the-use-of-the-msvc-compilers-on-windows) to use MSVC via a setup setting that will be ignored on non-windows. 
```
[tool.meson-python.args]
setup = ['--vsenv']
```

I think that is also the path to adding 64-bit interfaces via a `compile = ['-DBLAS_SYMBOL_SUFFIX=64_']` setting.

mattip commented:
I will rebase this off main to clear the conflict and push a commit to add the `tool.meson-python.args` settings.

mattip commented:
hmm, maybe the `compile = ['-DBLAS_SYMBOL_SUFFIX=64_']` option is too much: it will also use 64-bit interfaces on 32-bit builds.

mattip committed:
use MSVC and force 64-bit OpenBLAS interfaces

mattip committed:
drop rtools in wheel builds, do -DDBLAS_SYMBOL_SUFFIX=64_ differently

mattip committed:
typo, install pkg-config on windows

mattip committed:
set PKG_CONFIG_PATH for windows, add CXXFLAGS

mattip commented:
Getting somewhere. 

The manylinux wheel builds are failing with
```
FAILED core/tests/test_umath.py::TestComplexFunctions::test_loss_of_precision[complex64]
FAILED core/tests/test_umath.py::TestComplexFunctions::test_loss_of_precision[complex128]
FAILED core/tests/test_umath.py::TestComplexFunctions::test_loss_of_precision[clongdouble]
```

The musllinux builds have lots of DeprecationWarnings in tests, and fails `python /project/tools/wheels/check_license.py`

MacOS are failing some einsum tests
```
FAILED core/tests/test_einsum.py::TestEinsum::test_einsum_sums_int8 - Asserti...
FAILED core/tests/test_einsum.py::TestEinsum::test_einsum_sums_uint8 - Assert...
```

Windows is failing to find OpenBLAS, the `PKG_CONFIG_PATH` is probably not propagating properly.

PyPy does not build, it seems the include paths are not right?

mattip committed:
disable pypy builds, move PKG_CONFIG_PATH for windows

rgommers commented on pyproject.toml:
```diff
@@ -154,7 +147,7 @@ test-command = "bash {project}/tools/wheels/cibw_test_command.sh {project}"
 manylinux-aarch64-image = "manylinux2014"
 musllinux-x86_64-image = "musllinux_1_1"
-environment = { CFLAGS="-std=c99 -fno-strict-aliasing", LDFLAGS="-Wl,--strip-debug", OPENBLAS64_="/usr/local", NPY_USE_BLAS_ILP64="1", RUNNER_OS="Linux" }
+environment = { CFLAGS="-std=c99 -fno-strict-aliasing -DBLAS_SYMBOL_SUFFIX=64_ -DHAVE_BLAS_ILP64", LDFLAGS="-Wl,--strip-debug", CXXFLAGS="-DBLAS_SYMBOL_SUFFIX=64_ -DHAVE_BLAS_ILP64", OPENBLAS64_="/usr/local", NPY_USE_BLAS_ILP64="1", RUNNER_OS="Linux"}
```
This should be fine for now; we'll replace it with a `-Duse-ilp64` CLI flag in `meson_options.txt` later on.

rgommers commented:
> meson-python [can handle setting up windows](https://meson-python.readthedocs.io/en/latest/how-to-guides/meson-args.html#force-the-use-of-the-msvc-compilers-on-windows) to use MSVC via a setup setting that will be ignored on non-windows.

This seems a bit too agressive, it's better to make it specific to cibuildwheel to avoid forcing MSVC usage on everyone who builds from source. Probably this is the way:
```toml
[tool.cibuildwheel.windows]
config-settings = {"setup-args=--vsenv"}
```
or alternatively as an env var:
```
CIBW_CONFIG_SETTINGS_WINDOWS: "setup-args=--vsenv"
```

rgommers commented:
> PyPy does not build, it seems the include paths are not right?

I thought that was fixed for SciPy so it should work - good idea to disable that build config like you did and do that separately though.

> MacOS wheels are failing some einsum tests

That looks odd. Also there's a `test_new_policy` failure in the macOS jobs. 

It'd be useful to make the cibuildwheel jobs non-verbose, the CI logs are hard to navigate now - the log is 23,900 lines long. It's the ` verbose=2` usage in tools/wheels/cibw_test_command.sh`. Do you see a reason to keep them verbose?

> The musllinux wheel builds have lots of DeprecationWarnings in tests, and fails `python /project/tools/wheels/check_license.py`

I am unsure why other jobs are not failing on that, but the `LICENSE.txt` file is not installed into `site-packages/numpy` on purpose. It belongs in `.dist-info` not inside the source tree.
```
   + python /project/tools/wheels/check_license.py
  Traceback (most recent call last):
    File "/project/tools/wheels/check_license.py", line 54, in <module>
      main()
    File "/project/tools/wheels/check_license.py", line 38, in main
      with open(license_txt, encoding="utf-8") as f:
  FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmp.MGkClm/venv/lib/python3.10/site-packages/numpy/LICENSE.txt
```
So it's the test that needs tweaking here.



> The manylinux wheel builds are failing with
> 
> ```
> FAILED core/tests/test_umath.py::TestComplexFunctions::test_loss_of_precision[complex64]
> ```

Those seem to be gone, at least I checked two jobs and it only shows the same `test_new_policy` failure as on macOS.

mattip commented:
The wheels are building with [`meson setup ... -Dbuildtype=release -Db_ndebug=if-release`](https://github.com/numpy/numpy/actions/runs/5237346814/jobs/9455472404#step:5:323) which sets the compiler optimization to -O3`. I think this is the source of some of the test failures. The `-Ddbuildtype=release` is a meson-python default, where the non-wheel builds are using meson without meson-python. I think this should be a general default overrid for meson-python, we don't want users getting failed tests out of the box. At some point we should fix the failures.

rgommers commented:
@mattip do we want to default to `-O2`, like `distutils` does? Seems fine to me, I don't have a strong preference here. We actually have different defaults hardcoded per compiler in `numpy.distutils`, e.g. the Intel compilers are using `-O1`. Overall that's a bit of a mess, so I think having a single uniform default and then treating the rest as either numpy bugs or compiler bugs seems cleaner.

mattip commented:
Using 
```
[tool.meson-python.args]
setup = ['-Dbuildtype=custom','-Doptimization=0', '-Db_ndebug=true']
```

gives me
```
   User defined options
      Native files: /project/.mesonpy-fjiq3ch_/build/meson-python-native-file.ini
      buildtype   : custom
      optimization: 0
      b_ndebug    : true
      b_vscrt     : md
```

but the three tests still fail. I will commit what I have so far so others can try. I am not sure what is the delta to the other meson builds or to th distutil builds.

There are also DeprecationWarnings `Conversion of an array with ndim > 0 to a scalar is deprecated` in tests.

mattip committed:
fixes from review, try a -O0 build

rgommers commented:
> There is still a problem with the license file. I think the syntax here

Ah right, yes I had counted on PEP 639 moving forward by now. Let me have a look at this one right now.

rgommers committed:
BLD: handle license files inclusion in wheels correctly

rgommers commented:
> Windows is still not finding openblas. On the non-wheel meson builds, I used powershell to set up the openblas stuff, maybe reusing the bash script then setting the PKG_CONFIG_PATH elsewhere is not working right.

Did you look at how SciPy does this? Should work the same I'd think: https://github.com/scipy/scipy/blob/main/pyproject.toml#L163-L171

mattip commented:
I think I am doing the same as SciPy, but maybe I need to use `/` instead of `\\`.

There are still complex `test_loss_of_precision` failures on x86_64 that do not appear in the `linux_meson` CI job. Here is a theory: the lack of SIMD detection means that the wheel job, which is running in a manylinux2014 container, is using the older glibc/libm routines. The linux_meson CI job runs natively on Ubuntu 22.04, which has newer glibc/libm routines. The non-meson wheel jobs may not be using glibc/libm at all since they use SIMD code (is that true for the single-value checks in the test?).

The windows_meson (non-wheel) job is failing the test for an entry point for array api. Is there a problem with meson + windows + entry points?

rgommers commented:
> The windows_meson (non-wheel) job is failing the test for an entry point for array api. Is there a problem with meson + windows + entry points?

As we agreed before, I'll investigate that one (it's not just Windows, the test is failing on all platforms). If it's a blocker, we can put the `xfail` back in place and open a follow-up issue.

mattip committed:
windows fixes, xfail test on manylinux2014

rgommers commented on numpy/core/tests/test_umath.py:
```diff
@@ -4160,6 +4160,10 @@ def test_against_cmath(self):
                 assert_(abs(a - b) < atol, "%s %s: %s; cmath: %s" % (fname, p, a, b))
 
+    @pytest.mark.xfail(
+        _glibc_older_than("2.17"),
```
It is 2.17, so "older than" won't xfail the test, right?

mattip replied:
right, I was copying the pattern from elsewhere. I will update it to 2.18

mattip commented:
The non-wheel windows build is using a gcc from rtools43 to build NumPy. [These errors](https://dev.azure.com/numpy/numpy/_build/results?buildId=29875&view=logs&j=6f2fd294-e991-57ff-f3fe-958d72671852&t=11643d8d-7e48-55d1-bf9f-a0ee078a69cd&l=8364) seems to be fallout from that:
```
  ../../numpy/core/src/common/numpyos.c:515:5: error: unknown type name 'locale_t'; did you mean '_locale_t'?
    515 |     locale_t clocale;
        |     ^~~~~~~~
        |     _locale_t
  ../../numpy/core/src/common/numpyos.c:565:15: warning: implicit declaration of function 'newlocale'; did you mean 'setlocale'? [-Wimplicit-function-declaration]
    565 |     clocale = newlocale(LC_ALL_MASK, "C", NULL);
        |               ^~~~~~~~~
        |               setlocale
  ../../numpy/core/src/common/numpyos.c:565:25: error: 'LC_ALL_MASK' undeclared (first use in this function)
    565 |     clocale = newlocale(LC_ALL_MASK, "C", NULL);
        |                         ^~~~~~~~~~~
  ../../numpy/core/src/common/numpyos.c:565:25: note: each undeclared identifier is reported only once for each function it appears in
  ../../numpy/core/src/common/numpyos.c:568:18: warning: implicit declaration of function 'strtold_l'; did you mean 'strtoll'? [-Wimplicit-function-declaration]
    568 |         result = strtold_l(s, endptr, clocale);
        |                  ^~~~~~~~~
        |                  strtoll
  ../../numpy/core/src/common/numpyos.c:569:9: warning: implicit declaration of function 'freelocale'; did you mean '_free_locale'? [-Wimplicit-function-declaration]
   569 |         freelocale(clocale);
        |         _free_locale
```

I think we should go back to forcing a vsenv build on windows.

rgommers commented:
> I think we should go back to forcing a vsenv build on windows.

Yes, fine with me - I thought that was the plan indeed, given that we determined last week that we don't need `gfortran` anymore now that the openblas tarball comes with a pregenerated `.lib` file.

mattip committed:
fix PKG_CONFIG_PATH, force vsenv on windows builds, fix glibc version

mattip committed:
reuse parts of scipy's scripts for windows includding delvewheel; fix non-wheel build

mattip commented:
Progress. 
- The windows wheels are failing a [single test](https://github.com/numpy/numpy/actions/runs/5257920993/jobs/9501443535?pr=23838#step:5:1987), something about missing `npyrandom.lib` when compiling a cython extension module. I will xfail that test for now.
- The non-wheel windows builds are failing to test, something about the powershell syntax is off. I think I have a fix.
- The azure 32-bit linux build is failing [something in meson](https://dev.azure.com/numpy/numpy/_build/results?buildId=29883&view=logs&j=94356751-2a09-5f35-07c1-87bf12815f8e&t=aed0f7cb-c656-5ca0-8fcd-90f7c267ae7c&l=426), it says the size of double is 12 bytes.
- The cirrus CI tests are all failing. It seems to be something around [complex to string](https://cirrus-ci.com/task/5062656302776320?logs=cibuildwheel#L2524) round-tripping. Strange.

Now that we are skipping the test that led to trying a `-O0` build, let's see what happens if I revert that and we use the meson-python defaults.

mattip committed:
tweaks and xfail a failing test on windows

mattip commented:
52 CI runs are successful, 17 CI runs are still failing
#### github actions (7)
- 1 meson-windows run is newly failing after this PR, failing to find an api entry point in `test_array_api_entry_point`. This seems to be the only CI run where that happens, but note the azure windows runs have not gotten to the point where this test runs.
- 3 macos wheels runs are failing in [`TestEinsum::test_einsum_sums_int8`](https://github.com/numpy/numpy/actions/runs/5259928662/jobs/9506169142?pr=23838#step:5:2763). They were passing before [c607376](https://github.com/numpy/numpy/pull/23838/commits/c60737679ce262b8af5760641a9b41b936f00b1c), so maybe they now fail with `-O3`.
- 3 win32 wheel runs are failing to build with meson since the [host cpu and the python cpu differ](https://github.com/numpy/numpy/actions/runs/5259928662/jobs/9506170059?pr=23838#step:5:298)
#### azure (5)
- 3 win64 azure runs are failing to run tests via `python -m pytest numpy`. This used to use `runtests.py`, and needs some more care to avoid the "do not import numpy from the source directory" error. Maybe we should move these runs to github actions (which is also currently failing...).
- 1 win64 azure pypy run is failing to import numpy after building, perhaps the OpenBLAS DLL is not being properly copied. Maybe this too could move to github actions.
- 1 "failure" is a notice from azure that the build failed.
#### cirrusCI (5)
- 5 cirrusCI runs are failing as mentioned above.

#### circleCI
The run succeeds.

rgommers committed:
TST: Azure: avoid running tests from root of repo

This avoids the following failure mode:
```
ImportError while loading conftest 'D:\a\1\s\numpy\conftest.py'.
numpy\__init__.py:130: in <module>
    raise ImportError(msg) from e
E   ImportError: Error importing numpy: you should not try to import numpy from
E           its source directory; please exit the numpy source tree, and relaunch
E           your python interpreter from there.
```

rgommers committed:
TST: put back xfail for array-api entrypoint temporarily

rgommers committed:
REL: don't build 32-bit Python wheels with cibuildwheel

rgommers commented:
There are string repr/printing failures for `longdouble`. I'm going to mark these as xfail and file a follow-up issue. And post on the mailing list that folks who actually care about long double should start pitching in. It isn't worth our/my time to continue to investigate fiddly stuff like that.
```
FAILED tests/test_arrayprint.py::TestComplexArray::test_str - AssertionError: 
FAILED tests/test_longdouble.py::test_repr_roundtrip - AssertionError: 
FAILED tests/test_print.py::test_float_types[longdouble] - AssertionError: 
FAILED tests/test_print.py::test_nan_inf_float[longdouble] - AssertionError: 
FAILED tests/test_print.py::test_complex_types[clongdouble] - AssertionError: 
FAILED tests/test_print.py::test_complex_inf_nan[clongdouble] - AssertionError: 
FAILED tests/test_print.py::test_float_type_print[longdouble] - AssertionError: 
FAILED tests/test_print.py::test_complex_type_print[clongdouble] - AssertionE...
FAILED tests/test_scalarprint.py::TestRealScalars::test_str - AssertionError: 
FAILED tests/test_scalarprint.py::TestRealScalars::test_dragon4_interface - A...
FAILED tests/test_strings.py::test_float_to_string_cast[g-S] - AssertionError: 
FAILED tests/test_strings.py::test_float_to_string_cast[g-U] - AssertionError: 
12 failed, 19555 passed, 851 skipped, 17 xfailed, 3 xpassed, 20 warnings in 220.84s (0:03:40)
```

rgommers committed:
TST: disable longdouble string/print tests on Linux aarch64

See gh-23974

rgommers committed:
BLD: hack around broken long double size detection on 32-bit Linux

rgommers committed:
TST: fix deprecation warnings from f2py test for ndim>0

Representative subset of warnings observed in CI:
```
f2py/tests/test_return_integer.py: 20 warnings
  /tmp/tmp.roLIRtEhPN/venv/lib/python3.9/site-packages/numpy/f2py/tests/test_return_integer.py:24: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
    assert t(array([123], "d")) == 123

f2py/tests/test_return_real.py: 20 warnings
  /tmp/tmp.roLIRtEhPN/venv/lib/python3.9/site-packages/numpy/f2py/tests/test_return_real.py:23: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
    assert abs(t(array([234])) - 234.0) <= err

f2py/tests/test_return_real.py: 20 warnings
  /tmp/tmp.roLIRtEhPN/venv/lib/python3.9/site-packages/numpy/f2py/tests/test_return_real.py:24: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
    assert abs(t(array([[234]])) - 234.0) <= err
```

rgommers commented:
@mattip I finally looked into the `test_new_policy` failure, and there's nothing wrong with the memory allocator itself, but the test is written in a way that I think should be changed - recursively running the whole `numpy.core ` and `numpy.ma` test suite is super slow, confusing, and it will fail when any unrelated test in those modules fails. I think a better way of doing that is simply adding an env var or pytest mark to run those tests separately in a single CI job only.

rgommers committed:
TST: xfail einsum int8/uint8 test temporarily

mattip commented:
> I think a better way of doing that is simply adding an env var or pytest mark to run those tests separately in a single CI job only

Makes sense. I will open an issue.

rgommers commented:
I think it's also time to switch from `spin` to `dev.py`; the current workflow is still too cumbersome for effective debugging. For example:
- the equivalent of `python dev.py -s core` is: `spin test -- --pyargs numpy.core -m "not slow"` (xref https://github.com/scientific-python/spin/issues/40 for more examples)
- there is no equivalent of `python dev.py doc` other than manually setting PYTHONPATH and then running `make html` in `doc/`

rgommers commented:
For some reason it's extremely difficult to disable 32-bit Windows wheel building; I've used both `skip` (in two palces) and `archs` now without success. The `cibuildwheel` docs only have examples for Linux and macOS.

rgommers committed:
TST: disable `test_new_policy` test for memory allocator.

This is way too slow, running a large part of the test suite twice.
Issue 23975 tracks changing how this feature is tested.

mattip commented:
I think you can add [`CIBW_ARCHS_WINDOWS=auto64`](https://cibuildwheel.readthedocs.io/en/stable/options/#archs) around [here](https://github.com/numpy/numpy/blob/f52cb49037d1e51b26eeca16bd42844cc41d4519/.github/workflows/wheels.yml#L118) in the wheel.yml file

rgommers commented:
> I think you can add [`CIBW_ARCHS_WINDOWS=auto64`](https://cibuildwheel.readthedocs.io/en/stable/options/#archs) around [here](https://github.com/numpy/numpy/blob/f52cb49037d1e51b26eeca16bd42844cc41d4519/.github/workflows/wheels.yml#L118) in the wheel.yml file

Ah thanks, that explains it! The `skip`'s in `pyproject.toml` seem entirely unused then, the jobs are indeed controlled in `wheels.yml`.

rgommers commented:
The 32-bit long double detection hack isn't quite working: `undefined symbol: Dragon4_PrintFloat_Intel_extended128`. May have to punt on that.

rgommers committed:
CI: avoid use of pytest for 32-bit Linux Azure job; remove xml coverage

rgommers commented:
The macOS arm64 on Cirrus CI are still failing with:
```
_multiarray_umath.cpython-39-darwin.so, 0x0002): symbol not found in flat namespace (_cblas_caxpy)
```
That's related to forcing `NPY_USE_BLAS_ILP64="1"` but not using `-DBLAS_SYMBOL_SUFFIX=64_ -DHAVE_BLAS_ILP64` in `CFLAGS` unlike on GHA for x86-64.

Other than that, this is getting close. Trying to get it green by tonight.

rgommers committed:
CI: fix ILP64 usage on macOS arm64 with cibuildwheel

Also remove a no longer needed `-std=c99` (this is already specified
in the top-level `meson.build` file).

rgommers committed:
CI: disable pypy3.9 job on Azure temporarily

rgommers committed:
BLD: on 32-bit Linux, if `sizeof(long double)` is 12, change format

[skip circle] [skip cirrus]

rgommers commented:
Okay, this is ready to go in, phew:) All the follow-ups to this PR are listed in gh-23981.

mattip commented on .github/workflows/build_test.yml:
```diff
@@ -436,7 +436,7 @@ jobs:
         export CC=/usr/bin/gcc-12
         export CXX=/usr/bin/g++-12
-        python -m pip install -e .
+        python setup.py develop
```
Shouldn't this use meson?

rgommers replied:
This is a SIMD-related job, right? I left all of those on `setup.py` builds.

rgommers replied:
I'll resolve this; there are a number of CI jobs still to move over, this is simply one of them. It's tracked in gh-23981.

mattip commented on azure-pipelines.yml:
```diff
@@ -114,16 +114,9 @@ stages:
-            python3 -m pip install threadpoolctl && \
-            python3 tools/openblas_support.py --check_version"
+            cd tools && \
+            python3 -m pytest --pyargs numpy"
```
This has removed the check for the openblas version. Is that on purpose?

rgommers replied:
Yes, we now have proper logging in the build log for the version, and this cannot really be wrong (no other OpenBLAS version is installed in CI). This job was painful to debug, so I trimmed it down as much as possible.

mattip commented on numpy/random/tests/test_extending.py:
```diff
@@ -44,6 +46,7 @@
         cython = None
 
 
+@pytest.mark.xfail(IS_WINDOWS, reason="Wheel builds are missing npyrandom.lib")
```
I think this should be investigated in #23981

rgommers replied:
It's already listed there

mattip reviewed (approved):
LGTM, except for a few nits and checks for things that might need to be in the tracking issue

rgommers committed:
Merge remote-tracking branch 'upstream/main' into default-to-meson

rgommers commented:
This is still green after merging in the `long double` format detection fix from `main`. The commit history is clean enough I'd say and useful, so I'll prefer to merge it as is.

I'll go ahead and merge this to keep the ball rolling. Thanks for the reviews and help @mattip!

seberg commented on meson.build:
```diff
@@ -4,7 +4,7 @@ project(
   # It is dynamically added upon import by versioneer
   # See `numpy/__init__.py`
-  version: '1.24.0.dev0',
+  version: '2.0.0.dev0',
```
Seeing this here, do we have to liberate the runtime version for the wheel upload? to rename them?!

rgommers replied:
Liberate? I'm happy with the `2.0.0.dev0-cp39` etc wheel names.

mattip replied:
xref #24065. There is currently a problem uploading periodic builds: the upload script does not upload if the name alreeady exists. Two solutions, each with their own tradeoffs:
- upload anyway
- add a tag to the uploaded name

rgommers replied:
Just found it, thanks. Resolving here since we should keep the discussion on the issue. there's nothing wrong in this PR.
